plonky2/src/iop/challenger.rs

```diff
@@ -176,14 +180,23 @@ impl<F: RichField + Extendable<D>, H: AlgebraicHasher<F>, const D: usize>
         }
     }
 
+    pub fn from_state(sponge_state: [Target; SPONGE_WIDTH]) -> Self {
+        let output_buffer = sponge_state[0..SPONGE_RATE].to_vec();
```

Comment:
It seems like this assumes these duplex elements haven't already been consumed from the output buffer (before we saved and restored Challenger's state). That should be true with our current usage, since we call `duplex` before saving the state.

But still, it's not obvious why this is safe, and it could be used in an unsafe way if we're not careful. I'd suggest starting with an empty output buffer so there's no risk of accidentally reusing outputs.

I.e. we could replace `state()` with a `compact()` function which
- Processes any inputs (if none, we can skip `duplexing()`)
- Clears the output buffer
- Returns state

Reply:
Hm so if I change the `state()` function to the `compact()` function you suggested, I shouldn't need the modify this function `from_state()`, right?

Reply:
I was thinking just `output_buffer` would be set to `vec![]` in that case, but it could be the same otherwise